### ISSUE TITLE
How to get file exact file location (start and end positions) of all metadata in any audio file?

### ISSUE THREAD

**Issue by colinleet** [Zeugma440/atldotnet#90]
## My problem (… feature request!)

I'm having trouble figuring out a way to get the start and end file locations of the audio metadata for any particular type of track.   I need this information to determine what parts of the file I should exclude from a checksum of just the audio portion of the file -- to help me identify when a track's metadata has been externally edited.

## Environment

* ATL version: Most recent release (https://github.com/Zeugma440/atldotnet/commit/6f5282bee90b6b8dcf4d9187111b7a6fc40ac06b)

## Details

I'm currently using ATL to read all of the metadata in a music library and then populate an SQLite database with the information.  I'm basically done with this part but I'm struggling with a corner case of how to determine when a file has had it's metadata edited, or file moved (or renamed) without adding a doubled record of the track within the database. 

My temporary solution was to generate a checksum of the file, and then to use that as the primary uniqueness identifier (like how git does it) -- this allows me to move or rename the file without it being seen as a new file.  However I'm hitting a snag wherein if any of the metadata is edited, as that naturally edits the source file, it also scrambles the checksum... Mostly making my solution moot.

What I planned to do was to just generate a checksum based off of the start of the audio content of the file (first 512 kb or something like that)… However as different audio file types store their metadata in different locations of the file (start, end, middle for MP4, etc) in order to do this I need to know what portions of the file need to be excluded from the checksum.

After reading through a lot of the IO classes for individual audio file types it seems like each of them individually do keep track of this information (either through their own private headers for each file type, or through zone mappings which aren't used by all of the data types, or the information is just hardcoded in the headers)… But non of that info seems to be exposed publicly.

I was wondering if was any way of working around this to generate the information as to make it public / to add it to the info available within the interface IMetaDataIO?

Best Regards,
- Colin

**Comment by Zeugma440** [comment]
Thanks for the detailed explanation ! It's true that ATL records this data at some point or another during the processing. However, nothing has been streamlined or published so far as I didn't expect users to be interested in such low-level data.

Regarding your needs :
- Knowing that audio files can be tagged with **multiple** tagging formats, ATL can either provide multiple tag offset and sizes (one for each tagging format), or just the offset and size of audio data, which seems better imho
- Formats that are using MP4 tagging won't be easy to handle, as MP4 has metadata all over the place (depending on how the file is structured, chapter data chunks may be multiplexed with audio and video data chunks, just like subtitles). I guess that will be a limit to what you're trying to acheive
- The solution I have in mind will require to use lower level classes such as `AudioDataManager`, as I don't want to burden `Track` with such information. `Track` aims at providing high-level generic metadata.

**Comment by colinleet** [comment]
Thanks for the fast response!

"Knowing that audio files can be tagged with multiple tagging formats, ATL can either provide multiple tag offset and sizes (one for each tagging format), or just the offset and size of audio data, which seems better imho"

I agree, getting a single (or series for mp4 derivatives) of offsets + size of audio data would be easier and better for my situation.  Mostly I was just thinking that as the information was already being decoded for the opposite case, it would have been for me the most simple in terms of mining data already being generated.

"Formats that are using MP4 tagging won't be easy to handle, as MP4 has metadata all over the place (depending on how the file is structured, chapter data chunks may be multiplexed with audio and video data chunks, just like subtitles). I guess that will be a limit to what you're trying to acheive"

Yup again.  I'm just using (all) of the information which is in the "standard" set of metadata available within Track(), add that's what I designed my database around.  I'm not really interested in reading chapter markers / lyrics and subtitles, mp4 is more of a corner case for when someone is trying to play a download music video.

"The solution I have in mind will require to use lower level classes such as AudioDataManager, as I don't want to burden Track with such information. Track aims at providing high-level generic metadata."

That makes sense and is good design for keeping the data mode compartmentalized.  Keeping the data within the lower level class which directly deals with the audio file is better: just as long as the info is pubic and accessible with the same methods or accessories across the different audio file type, and the lower class level is publicly/simply callable like how Tracks(audioFilePath) is.

What solution do you have in mind?

**Comment by colinleet** [comment]
I can deal with not as abstracted levels of the file type readers as well. I was just saying what would be the easier for me.

**Comment by Zeugma440** [comment]
Right now, what I have in mind is something like

```csharp
AudioDataManager theFile = new AudioDataManager("c:/audio/test.mp3");
theFile.ReadFromFile();
long audioOffset = theFile.AudioDataOffset;
long audioSize = theFile.AudioDataSize;
```

I'm taking some time as there are a lot of supported formats already, and each of them has its own specifics 😉

**Comment by colinleet** [comment]
Perfect! Thank you!

**Comment by Zeugma440** [comment]
I just committed a big change on `master` that should do what you want, according to the above code sample.
Would you like to test it as is ?

**Comment by colinleet** [comment]
I implemented the new code and it's returning the same checksum after I edit the metadata!  Success!
[Note: I only really tested it on 13 mp3s so it's not a great universal test.]

Now I'll need to fix some bugs in my code where it's not catching the duplicate MD5s (but that's all on my end)....

There was just one thing that I think need to be cleaned-up, my code to run AudioDataManager:

'''
        /// <summary>
        /// Returns the start index of the audio data within the file and the length of the audio data.
        /// </summary>
        /// <param name="fileName">Audio file to get the start and length of the audio chunk for.</param>
        private static (long, long) TrackAudioStartLocation(string fileName) {
            byte alternate = 0;
            IProgress<float> writeProgress = null;

            // Try to get the position using the first factory.
            var audioData = AudioDataIOFactory.GetInstance().GetFromPath(fileName, alternate);
            var audioManager = new AudioDataManager(audioData, writeProgress);
            bool found = audioManager.ReadFromFile();

            // Try alternate factories till the track is found.
            while ( !found && alternate < AudioDataIOFactory.MAX_ALTERNATES ) {
                alternate++;
                audioData = AudioDataIOFactory.GetInstance().GetFromPath(fileName, alternate);
                audioManager = new AudioDataManager(audioData, writeProgress);
                found = audioManager.ReadFromFile();
            }

            // Return the position in the file.
            if ( found ) {
                return (audioManager.AudioDataOffset, audioManager.AudioDataSize);
            } else {
                return (-1, -1);
            }
        }
'''

Problems 1 ) There still isn't a constructor for AudioDataManager which just takes in a audio path.
The AudioDataManager still requires a IAudioDataIO to be passed into to when creating an instance of it, which change with each audio format.

2) Currently the code to use alternate metadata writing styles is bundled up in  AudioFileIO -- which is an internal class is not accessible without a bit of work ( basically what is above ).

Additionally the private AudioDataManager within each Track() is still private... I would suggest just exposing that one with a read only accessor to eliminate needing to add the alternate path-only-constructor for AudioDataManager (like your example above showed).... Or you could just create one similar to the function I added here.

Thank you!

**Comment by Zeugma440** [comment]
The code you've written is completely fine, but I agree it's an unnecessary complexity to handle on your side.

However, having to choose between making an `internal` class `public` and adding "low-level" fields to `Track`, I happily choose the latter.

=> I've created `Track.TechnicalInformation` (and also made all setters for physical properties `internal` for clarity's sake, as ATL cannot be used to alter sample rates, bitrates and other non-meta info that is published by `Track`).

Is that better on your side ?

**Comment by colinleet** [comment]
Perfect! Thank you.

**Comment by Zeugma440** [comment]
I've made the changes available in today's v3.20. Enjoy~
